Allows for Segwit functionality

bjarnemagnussen opened the pull request:
I have been working on implementing Segwit into Bit and now that the multisig-part was merged I want to release this code as a pull request.

This pull request adds a lot of code and also changes some of the multisig part slightly to make it more readable. But the evolution of the code should be relatively straight forward from the commits.

The new features include:

 * Send to P2SH and Bech32 outputs.
 * Transaction batching (combining inputs spending from different keys' unspents).
 * Generating P2SH-nested Segwit addresses for both single keys (PrivateKey) and multisig (MultiSig).

This pull request does not include the possibility to use Bech32 address to receive bitcoins. It would be straight-forward to implement, but Bit depends on external blockexplorer APIs that have none or only very limited Bech32 support.

Examples are provided below. 

Tests have been added accordingly, but maybe someone can come up with better test methods. Also the way segwit addresses for the keys are generated may be a bit cumbersome and at some point needs refactorizing the code a little. 

Suggestions for improvements are welcome! Please let me know if you find any errors while testing the code.

### Initialising the code
```
from bit import *
from bit.transaction import *
from bit.network.meta import *
```
### Creating Segwit transaction using unspents from both legacy and segwit outputs
```
key1 = PrivateKeyTestnet("KzDFE5K1Mb6cP2SAUySLUMx8F8KWn5RYiUFTFXcDMY5X22Jp1MvH")
key1.address
#'mtgECN1DaFVvqrGGdF517prMBSzC2AkVL1'
key1.segwit_address
#"2N21Gzex7WJCzzsA5D33nofcnm1dYSKuJzT"

# Unspents from key1 (could also be retrieved using key1.get_unspents() or calling key1.create_transactions from below without specifying any unspents):
unspents_key1 = [
	Unspent(100000000, 1, '76a914905aa72f3d1747094a24d3adbc38905bb451ffc888ac', 'f09c22717770fcd7e477953c3ca7ecb9bd44ec4d5392f24fdd247dbb8db2d388', 0, False),
	Unspent(3000000000,1 ,'a9146015d175e191e6e5b99211e3ffc6ea7658cb051a87', '37af0739136575bffa08a35269a360c047a06e5816b74383f187802c09d7dee2', 1, True)]

# Send 0.5 bitcoins to  address 2NEcbT1xeB7488HqpmXeC2u5zqYFQ5n4x5Q:
key1.create_transaction(outputs=[("2NEcbT1xeB7488HqpmXeC2u5zqYFQ5n4x5Q", 50000000, "satoshi")], fee=200, unspents=unspents_key1)
#0100000000010288d3b28dbb7d24dd4ff292534dec44bdb9eca73c3c9577e4d7fc707771229cf0000000006a4730440220750542b1727b3774a7095eb12a20357e75d2ba51b5de13109780012588e0c3b2022032a38b0f9a186b69ea700492f73d7acb57670ac4e700c239b9b75932b2b27f700121021816325d19fd34fd87a039e83e35fc9de3c9de64a501a6684b9bf9946364fbb7ffffffffe2ded7092c8087f18343b716586ea047c060a36952a308fabf7565133907af370100000017160014905aa72f3d1747094a24d3adbc38905bb451ffc8ffffffff0280f0fa020000000017a914ea654a94b18eb41ce290c135cccf9f348e7856a287502acab50000000017a9146015d175e191e6e5b99211e3ffc6ea7658cb051a870002483045022100a62bd04b47c9a2046225303deed38cbb9bd9bc6cc3a9980e82a4826ad001df670220031ad40aad402d4d71f00229daac80ad48cdfcb298f0b27e5b45414947b6c82d0121021816325d19fd34fd87a039e83e35fc9de3c9de64a501a6684b9bf9946364fbb700000000
```
### Creating Segwit transaction using unspents from different keys (`PrivateKey` and `MultiSig`)
```
key2 = PrivateKeyTestnet("KxhDjpU1TWDbD2ukWpTcj9T55bkKVJsyn2hcDucCecpjApetCRG9")
p = [key1.public_key.hex(), key2.public_key.hex()]
multi1 = MultiSigTestnet(key1, p, 2)
multi1.segwit_address
#'2NCWeVbWmaUp92dSFP3RddPk6r3GTd6cDd6'
multi2 = MultiSigTestnet(key2, p, 2)

unspents_batch = [
    Unspent(200000000,
            1,
            'a914d35515db546bb040e61651150343a218c87b471e87',
            '64064449fc2aab13316e90beb6c2a224d86a08733ff5b48e422be7688de72346',
            0, True),
    Unspent(3000000000,
			 1,
			  'a9146015d175e191e6e5b99211e3ffc6ea7658cb051a87',
			  '37af0739136575bffa08a35269a360c047a06e5816b74383f187802c09d7dee2',
			  1,
			   True)
]

# multi2 creates the transaction and provides a partially-signed scriptsig to the input spending from an output corresponding to multi2's address:
tx1 = multi2.create_transaction(outputs=[("bcrt1qpm3x3jrdqhefptw3hlymmlpejttct08zgzzd2t", 400000000, "satoshi")], fee=220, unspents=unspents_batch)
tx1
#010000000001024623e78d68e72b428eb4f53f73086ad824a2c2b6be906e3113ab2afc494406640000000023220020d3a4db6921782f78eb4f158f73adde471629dd5aca41c14a5bfc2ec2a8f39202ffffffffe2ded7092c8087f18343b716586ea047c060a36952a308fabf7565133907af370100000000ffffffff020084d717000000001600140ee268c86d05f290add1bfc9bdfc3992d785bce2985ae3a60000000017a914d35515db546bb040e61651150343a218c87b471e87040047304402201b393cb5783ebc432cdb97163a8082f2c6490ebea3257dac2227a793ece136cb022002530c83b3a92a8b8a1f31db68d6d099c7adce01bb46b96dcc799af8fd5c9d4a0100475221021816325d19fd34fd87a039e83e35fc9de3c9de64a501a6684b9bf9946364fbb721037d696886864509ed63044d8f1bcd53b8def1247bd2bbe056ff81b23e8c09280f52ae0000000000

# key1 can sign its transaction input using only unspents from key1 (but we could also have provided the whole unspent-list):
tx1_1 = key1.sign_transaction(tx1, unspents=unspents_key1)
tx1_1
#010000000001024623e78d68e72b428eb4f53f73086ad824a2c2b6be906e3113ab2afc494406640000000023220020d3a4db6921782f78eb4f158f73adde471629dd5aca41c14a5bfc2ec2a8f39202ffffffffe2ded7092c8087f18343b716586ea047c060a36952a308fabf7565133907af370100000017160014905aa72f3d1747094a24d3adbc38905bb451ffc8ffffffff020084d717000000001600140ee268c86d05f290add1bfc9bdfc3992d785bce2985ae3a60000000017a914d35515db546bb040e61651150343a218c87b471e87040047304402201b393cb5783ebc432cdb97163a8082f2c6490ebea3257dac2227a793ece136cb022002530c83b3a92a8b8a1f31db68d6d099c7adce01bb46b96dcc799af8fd5c9d4a0100475221021816325d19fd34fd87a039e83e35fc9de3c9de64a501a6684b9bf9946364fbb721037d696886864509ed63044d8f1bcd53b8def1247bd2bbe056ff81b23e8c09280f52ae02483045022100d30b3caf525c0dd9e6837a4f4ef966cc9c34e40c59a4ae79e28305a3f7d93e03022026af047648a412be92171c07b4d706e3eaf91eb427b36e1a87dff58a3c2762920121021816325d19fd34fd87a039e83e35fc9de3c9de64a501a6684b9bf9946364fbb700000000
```
### Continuing example above: The order of the provided unspents does ***not*** have to correspond to the order of the inputs in the transaction
```
# Lastly multi1 can provde the missing signature to the partially-signed input and the order in which unspents are provided does not matter:
tx_signed = multi1.sign_transaction(tx1_1, unspents=unspents_batch[::-1])
tx_signed
#010000000001024623e78d68e72b428eb4f53f73086ad824a2c2b6be906e3113ab2afc494406640000000023220020d3a4db6921782f78eb4f158f73adde471629dd5aca41c14a5bfc2ec2a8f39202ffffffffe2ded7092c8087f18343b716586ea047c060a36952a308fabf7565133907af370100000017160014905aa72f3d1747094a24d3adbc38905bb451ffc8ffffffff020084d717000000001600140ee268c86d05f290add1bfc9bdfc3992d785bce2985ae3a60000000017a914d35515db546bb040e61651150343a218c87b471e870400483045022100a2e1d3c0f692c6aa31e659676896b333adf2b2ef081cc17b0166977adb287b2102207a2a8d69b20c290f64bd9abc93fb454a55bfd51f71866ede4d98aafd783e10ea0147304402201b393cb5783ebc432cdb97163a8082f2c6490ebea3257dac2227a793ece136cb022002530c83b3a92a8b8a1f31db68d6d099c7adce01bb46b96dcc799af8fd5c9d4a01475221021816325d19fd34fd87a039e83e35fc9de3c9de64a501a6684b9bf9946364fbb721037d696886864509ed63044d8f1bcd53b8def1247bd2bbe056ff81b23e8c09280f52ae02483045022100d30b3caf525c0dd9e6837a4f4ef966cc9c34e40c59a4ae79e28305a3f7d93e03022026af047648a412be92171c07b4d706e3eaf91eb427b36e1a87dff58a3c2762920121021816325d19fd34fd87a039e83e35fc9de3c9de64a501a6684b9bf9946364fbb700000000
```

teran-mckinney commented:
Looking good at first glance. Need more time to review, however. Is there no bech32 pip installable?

bjarnemagnussen commented:
I don't know of any bech32 pip installable library to use. Maybe we can find one. Sipa's reference code does the job well.

I have added commits which among others improve the fee estimation, see #44 . The fix adds to the Unspents-objects additional fields to denote the type (p2pkh/p2sh/np2wkh, etc.). 

Further, regarding #30 I have implemented the coin selection method "Branch and Bound" to try to find a perfect match when combine=False (with no return address) and otherwise falls back to a single random draw with return address with a threshold for a minimum value for the return address if desired. It uses random exploration, but it may be better to use largest first exploration, see discussion here:
https://github.com/bitcoin/bitcoin/pull/10637
For Bit I don't think it makes a big difference (using largest first exploration would actually be easier). I have not tested the performance extensively using uge unspent-sets, which should still be done.

Then I have added/changed some of the tests to make them more specific.

Lastly there is still one small thing missing, which is a correct tx-id calculation. With segwit the current calculate_txid() function will calculate the hash for segwit transactions, but not what is now considered the txid. As far as I can see the function is only called to display the  id when sending a transaction. It would probably make sense to refactor the code for calculating the transaction id to the TxObj instead. 

If you have any questions or suggestions for improvement let me know after you have had the time to take a look at the pull request.

cal5k commented:
@ofek Any chance you could merge this puppy in?

teran-mckinney commented:
Hey! Sorry for the wait. Please ping me if I'm ever too slow on stuff like this.

My main question is around Brand and Bound, if that is required for segwit or not. It sounds like a cool change but could have pretty big implications. Will combine=False/True still work? I skimmed some details like that. Ideally, PRs should be smaller but I know stuff like this is hard to keep to a minimum.

That's my only concern at the moment. @ofek, if the Brand and Bound stuff is okay, this checks out to me. Are you comfortable with me merging it?

Thanks!

teran-mckinney commented:
Thanks for getting back to me.

Unspent changes look good to me, no real comments there. Seems sensible to me although I don't have the most insight there.

Will give this a while longer, if no word from @ofek I will merge it in. I think we should do a release soon after.

ofek commented:
Sorry for the delay everyone.

@bjarnemagnussen Excellent work!

Let's merge this. Though before we release, please (in separate PRs):

1. Update docs
2. Get all tests passing

@teran-mckinney Merge away!
